### ISSUE TITLE
Bicycle network preference

### ISSUE THREAD

**Issue by wieringen** [valhalla/valhalla#3674]
In the Netherlands, there are different types of bicycle road networks. In OSM these different road networks are tagged with the tag: [network](https://wiki.openstreetmap.org/wiki/Tag:network%3Drcn). Some networks are designed to be functional and others try to provide more scenic routes.

Is there a way to prefer certain networks over others? Ideally, I would like to give more weight to ways tagged with "network=rcn". Is this something I can achieve with costing_options? If not can someone give me some pointers on how I can go about implementing this myself?

**Comment by nilsnolde** [comment]
the good news is: we already parse the `network` tag already. it's even capturing all possible values it seems (which is really important in this case, thanks to the team:)), but in the end, while doing the edge costing, it's [degrading that info to boolean](https://github.com/valhalla/valhalla/blob/32abe4f44169bc5db5a20b3fa1d1d6cd397671be/src/sif/bicyclecost.cc#L667-L670). it'd need another costing option in the request parameters to specify the kind of bike network you'd like (or even just "yes" for any network) and maybe even adjusting the costing for it a little? 0.95 seems very little. cc @dnesbitt61

**Comment by dnesbitt61** [comment]
bicycle network tags are not stored in the Valhalla directed edges. They store a boolean value that states whether a directed edge is part of a bicycle network is stored. That means there is no easy way to modify costing to favor specific bicycle networks over other bicycle networks. The actual network type is stored in EdgeInfo, but this is not usable in costing. One would have to modify the Valhalla tile structure to add specific bicycle network types to the DirectedEdge structure and then add costing options to the request to vary costing by the different network types.

The current costing that favors bicycle networks slightly (0.95) is mostly in place since in some parts of the US (at least where I cycle) the bicycle networks are on roads that I don't consider great for cycling. Perhaps in Western Europe these tags better represent preferred cycling roads/paths.

**Comment by kevinkreiser** [comment]
currently, the costing functions *do* get access to the tile though. motor vehicle costings use this tile access to look up realtime/historical speed information for an edge. anyway, because of this, it is also technically possible to use the edge to look in the tile to pull out the edgeinfo for the edge and then to find the exact bike network. it may not be advisable, but it *is* possible. perhaps because of the bit on the edge, we can avoid any significant penalty by only checking the edgeinfo->bike_network when we know its going to actually be there? anyway it would be something someone could experiment with if they were so inclined!

**Comment by dnesbitt61** [comment]
good point. I didn't mention this since bicycle routing can already be slow (for routes over 100miles or so) since there are no real optimizations apart from the A* heuristic. Generally it is inadvisable to access EdgeInfo within path finding since these records are not sequential, leading to poor memory coherence. Your point of checking if the bike_network flag is true prior to checking EdgeInfo for the actual network type would help keep the performance drop in check. Looking at Belgium, Netherlands, and some other countries the route networks are well populated so this would slow performance more than in areas like US.

**Comment by kevinkreiser** [comment]
yep! its probably not advisable but it is something that one could look into was all i wanted to point out. in such a scenario i would also imagine that we'd have a costing option to allow a person to opt into caring what type of bike network was desirable to them which would be a further bit we could check to further avoid looking at the edgeinfo unless they really really cared about it in the request.

**Comment by nilsnolde** [comment]
I like that idea, it's pretty low effort and gets the job done, albeit lower performance. maybe there should be a costing option **and** a config opt-in option (defaulting to false and conserving current behavior)? I guess if that's an option, then the costing factor should be also decreased when a segment of the specified bike network is traversed, like 0.2ish or even lower? to me, if a user explicitly specifies that, that person really wants to ride that network.

**Comment by dnesbitt61** [comment]
cost factors cannot effectively be as low as what you are saying - that will likely interact with the A* heuristic and lead to the possibility of very strange paths. It is better to penalize roads/paths that are not part of the cycling network if you want to have that much influence.

**Comment by nilsnolde** [comment]
thanks, got your point. just had a look at a few other low cost factors like `cyclelane_factor`:

https://github.com/valhalla/valhalla/blob/master/src/sif/bicyclecost.cc#L489-L496

> if you want to have that much influence.

well, I'm not confident if that's the right way to go, wanted to hear what other people think.

**Comment by wieringen** [comment]
Thanks for all the information guys. C++ is not my specialty, but I think I can make it work.

**Comment by davidmurray** [comment]
Hi @wieringen, I'm also looking to do something similar to this. Just wondering if you had some success in making this work!
Thanks!

**Comment by wieringen** [comment]
@davidmurray I'm still experimenting, so for now I did a quick fix and removed the other bike networks from the bike_network_mask that is defined in lua/graph.lua. This seems to work pretty well.

**Comment by farfromrefug** [comment]
@nilsnolde @dnesbitt61 @kevinkreiser i come back to that issue as i have a lot of demands on routing to favor bicycle networks  (could be out of road paths or road ones).
I read a bit through the discussion and i am not sure to understand what would be the best way to go. As i understand from @dnesbitt61 it would be best to penalize "non network" edges?
For now my need is not for per network factor. So would a else to [this](https://github.com/valhalla/valhalla/blob/master/src/sif/bicyclecost.cc#L667-L670) if case be enough to apply a factor to other edges?

**Comment by nilsnolde** [comment]
To patch it real quick: yes that’s fine. You can also simply replace the „if“, 0.95 isn’t doing much as a factor anyways, and in the end your „else“ would do the exact same thing. It’ll trigger much more often, so there’s a slight penalty to pay, which is probably the reason why it is the way it is.

**Comment by nilsnolde** [comment]
Seems ok for the case where you want the user to control it. Not sure the concept is ok for master, but it should work for you want to do.

> btw any chance we could add the "hike_network" for pedestrian?

Doesn't hurt to put a feature request for that.

**Comment by farfromrefug** [comment]
@nilsnolde how i can change it for master? I would like to create a PR for everyone to enjoy this

**Comment by nilsnolde** [comment]
If you want to contribute this, I'd say first raise a PR and then we can discuss there. IMO it's fine conceptually, but @dnesbitt61 is more the gatekeeper for bike stuff, so we'll need his review/opinion here. 

It's always a bit tricky with these avoiding things rather than preferring things. Although it's generally more powerful, it's harder to convey to the user what it does. A `non_network_factor` sounds terrible to be honest;) However, it's ok to invert a `bike_network_factor` (e.g. 0 - 1 with 1 really wanting to ride a network and 0, no change in costing), then in code we invert the meaning and use that inversion to penalize non-network edges. We did smth like in https://github.com/valhalla/valhalla/pull/3957.

Anyways, before doing more on it, PR what you have (and make sure to base your branch off _our_ master, not nutiteq's which is super old)
